src/pip/_internal/req/__init__.py

```diff
@@ -60,39 +70,118 @@ def install_given_reqs(
+            executed_parallel_reqs = map_multiprocess_ordered(
+                partial(_single_install, install_args, suppress_exception=True),
+                should_parallel_values
+            )
+            # collect back results
+            parallel_install_results = dict(
+                zip(should_parallel_indexes, executed_parallel_reqs))
+        else:
+            parallel_install_results = {}
+
+        # check the results from the parallel installation,
+        # and fill-in missing installations or raise exception
+        for i, req in enumerate(requirements):
+
+            # select the install result from the parallel installation
+            # or install serially now
             try:
-                requirement.install(
-                    install_options,
-                    global_options,
-                    root=root,
-                    home=home,
-                    prefix=prefix,
-                    warn_script_location=warn_script_location,
-                    use_user_site=use_user_site,
-                    pycompile=pycompile,
-                )
-            except Exception:
-                # if install did not succeed, rollback previous uninstall
-                if uninstalled_pathset and not requirement.install_succeeded:
-                    uninstalled_pathset.rollback()
-                raise
-            else:
-                if uninstalled_pathset and requirement.install_succeeded:
-                    uninstalled_pathset.commit()
-
-            installed.append(InstallationResult(req_name))
+                install_result = parallel_install_results[i]
+                if isinstance(install_result, InstallationResult):
+                    logger.debug(
+                        'Successfully installed %s in parallel', req.name)
+            except KeyError:
+                install_result = _single_install(
+                    install_args, req, suppress_exception=False)
+                logger.debug('Successfully installed %s serially', req.name)
+
+            # Now processes the installation result,
+            # throw exception or add into installed packages
+            if isinstance(install_result, BaseException):
+                # Raise an exception if we caught one
+                # during the parallel installation
+                raise install_result
+            elif isinstance(install_result, InstallationResult):
+                installed.append(install_result)
 
     return installed
+
+
+def _single_install(
+        install_args,                # type: _InstallArgs
+        requirement,                 # type: InstallRequirement
+        suppress_exception=False,    # type: bool
+):
+    # type: (...) -> Union[None, InstallationResult, BaseException]
+    """
+    Install a single requirement, returns InstallationResult
+    (to be called per requirement, either in parallel or serially).
+    Notice the two lists are of the same length
```

Comment:
It's not clear to me what lists are being referenced here?

Reply:
Leftover from a previous iteration 🙂 Need to fix this!

Reply:
👍